File: android-core/src/test/java/com/mparticle/identity/AliasRequestTest.java
package com.mparticle.identity;

import com.mparticle.MParticle;
import com.mparticle.MockMParticle;

import org.junit.Test;
import org.mockito.Mockito;

import java.util.Random;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class AliasRequestTest {
    Random random = new Random();

    @Test
    public void testAliasRequestBuilder() {
        MParticleUser sourceUser = Mockito.mock(MParticleUser.class);
        MParticleUser destinationUser = Mockito.mock(MParticleUser.class);

        long startTime = System.currentTimeMillis() - 100;
        long endTime = System.currentTimeMillis();
        long sourceMpid = random.nextLong();
        long destinationMpid = random.nextLong();

        Mockito.when(sourceUser.getId()).thenReturn(sourceMpid);
        Mockito.when(sourceUser.getFirstSeenTime()).thenReturn(startTime);
        Mockito.when(sourceUser.getLastSeenTime()).thenReturn(endTime);

        Mockito.when(destinationUser.getId()).thenReturn(destinationMpid);

        AliasRequest request = AliasRequest.builder(sourceUser, destinationUser).build();

        assertEquals(destinationMpid, request.getDestinationMpid());
        assertEquals(sourceMpid, request.getSourceMpid());
        assertEquals(request.getStartTime(), startTime);
        assertEquals(request.getEndTime(), endTime);
    }

    @Test
    public void testBasicBuilder() {
        long startTime = random.nextLong();
        long endTime = random.nextLong();
        long sourceMpid = random.nextLong();
        long destinationMpid = random.nextLong();

        AliasRequest.Builder  builder = AliasRequest.builder()
                .destinationMpid(destinationMpid)
                .sourceMpid(sourceMpid)
                .startTime(startTime)
                .endTime(endTime);

        AliasRequest request = builder.build();

        assertEquals(sourceMpid, request.getSourceMpid());
        assertEquals(destinationMpid, request.getDestinationMpid());
        assertEquals(startTime, request.getStartTime());
        assertEquals(endTime, request.getEndTime());
    }

    @Test
    public void testMaxWindowEnforcement() {
        MParticle.setInstance(new MockMParticle());
        Mockito.when(MParticle.getInstance().Internal().getConfigManager().getAliasMaxWindow()).thenReturn(1);

        MParticleUser sourceUser = Mockito.mock(MParticleUser.class);
        MParticleUser destinationUser = Mockito.mock(MParticleUser.class);

        long earliestLegalStartTime = System.currentTimeMillis() - (24 * 60 * 60 * 1000);

        long startTime = 0;
        long endTime = System.currentTimeMillis();
        long sourceMpid = random.nextLong();
        long destinationMpid = random.nextLong();

        Mockito.when(sourceUser.getId()).thenReturn(sourceMpid);
        Mockito.when(sourceUser.getFirstSeenTime()).thenReturn(startTime);
        Mockito.when(sourceUser.getLastSeenTime()).thenReturn(endTime);
        Mockito.when(destinationUser.getId()).thenReturn(destinationMpid);

        AliasRequest request = AliasRequest.builder(sourceUser, destinationUser).build();

        assertEquals(sourceMpid, request.getSourceMpid());
        assertEquals(destinationMpid, request.getDestinationMpid());

        //since request.getStartTime generates it's own earliestLegalStartTime based on System.currentTimeMillis(),
        //we need to give some wiggle room for time elapsed during test
        assertTrue(Math.abs(earliestLegalStartTime - request.getStartTime()) < 50);
        assertEquals(endTime, request.getEndTime());

    }
}

File: android-core/src/test/kotlin/com/mparticle/identity/AliasRequestTest.kt
package com.mparticle.identity

import com.mparticle.MParticle
import com.mparticle.MockMParticle
import org.junit.Assert
import org.junit.Test
import org.mockito.Mockito
import java.util.Random
import kotlin.math.abs

class AliasRequestTest {
    private var random = Random()

    @Test
    fun testAliasRequestBuilder() {
        val sourceUser = Mockito.mock(MParticleUser::class.java)
        val destinationUser = Mockito.mock(MParticleUser::class.java)
        val startTime = System.currentTimeMillis() - 100
        val endTime = System.currentTimeMillis()
        val sourceMpid = random.nextLong()
        val destinationMpid = random.nextLong()
        Mockito.`when`(sourceUser.id).thenReturn(sourceMpid)
        Mockito.`when`(sourceUser.firstSeenTime).thenReturn(startTime)
        Mockito.`when`(sourceUser.lastSeenTime).thenReturn(endTime)
        Mockito.`when`(destinationUser.id).thenReturn(destinationMpid)
        val request = AliasRequest.builder(sourceUser, destinationUser).build()
        Assert.assertEquals(destinationMpid, request.destinationMpid)
        Assert.assertEquals(sourceMpid, request.sourceMpid)
        Assert.assertEquals(request.startTime, startTime)
        Assert.assertEquals(request.endTime, endTime)
    }

    @Test
    fun testBasicBuilder() {
        val startTime = random.nextLong()
        val endTime = random.nextLong()
        val sourceMpid = random.nextLong()
        val destinationMpid = random.nextLong()
        val builder = AliasRequest.builder()
            .destinationMpid(destinationMpid)
            .sourceMpid(sourceMpid)
            .startTime(startTime)
            .endTime(endTime)
        val request = builder.build()
        Assert.assertEquals(sourceMpid, request.sourceMpid)
        Assert.assertEquals(destinationMpid, request.destinationMpid)
        Assert.assertEquals(startTime, request.startTime)
        Assert.assertEquals(endTime, request.endTime)
    }

    @Test
    fun testMaxWindowEnforcement() {
        MParticle.setInstance(MockMParticle())
        Mockito.`when`(MParticle.getInstance()?.Internal()?.configManager?.aliasMaxWindow)
            .thenReturn(1)
        val sourceUser = Mockito.mock(MParticleUser::class.java)
        val destinationUser = Mockito.mock(MParticleUser::class.java)
        val earliestLegalStartTime = System.currentTimeMillis() - 24 * 60 * 60 * 1000
        val startTime: Long = 0
        val endTime = System.currentTimeMillis()
        val sourceMpid = random.nextLong()
        val destinationMpid = random.nextLong()
        Mockito.`when`(sourceUser.id).thenReturn(sourceMpid)
        Mockito.`when`(sourceUser.firstSeenTime).thenReturn(startTime)
        Mockito.`when`(sourceUser.lastSeenTime).thenReturn(endTime)
        Mockito.`when`(destinationUser.id).thenReturn(destinationMpid)
        val request = AliasRequest.builder(sourceUser, destinationUser).build()
        Assert.assertEquals(sourceMpid, request.sourceMpid)
        Assert.assertEquals(destinationMpid, request.destinationMpid)

        // since request.getStartTime generates it's own earliestLegalStartTime based on System.currentTimeMillis(),
        // we need to give some wiggle room for time elapsed during test
        Assert.assertTrue(abs(earliestLegalStartTime - request.startTime) < 50)
        Assert.assertEquals(endTime, request.endTime)
    }
}
